### ISSUE TITLE
Reprocuding the MNIST results.

### ISSUE THREAD

**Issue by cfpark00** [DimaKrotov/Biological_Learning#4]
Hi, and thank you for this very interesting research.

I did run the given code and it does generate the expected hidden layer weights.

However, I can't reproduce the mnist accuracy(anything not mentioned is as the jupyter notebook):
training set:60000
test set: 10000

I tried with cross entropy to train the supervised part, and got a test-set accuracy of 92.68 %.

I tried the loss your paper mentions, but it gave worse results than CE for me.
Also, I wasn't able to use n=4.5 since that was giving nans in eq [1].

Do you have code to reproduce the results?

Thanks!

**Comment by DimaKrotov** [comment]
Hi, sorry I have just noticed this message. When plugging the inputs into equation [1], it is important to normalize them first, so that each MNIST image is a unit vector. Since the weights are also unit vectors, bc of the projector in the learning rule, the dot product of the two should be a cos-similarity, which is a number between -1 and +1. Thus, there should be no NaNs. This should bring the accuracy up to 98.5% approximately. It shouldn't matter much which loss you are using for training the classifier: both cross-entropy and the loss from the paper should lead to approximately similar accuracies. 

Also, please notice that it is not necessary to normalize the training data to be unit vectors for the unsupervised learning part, since the ranking operation used in the learning rule is invariant under multiplication by this normalization constant. But, if you want, you can normalize the data for this unsupervised part too (this will not change the results if the learning rate is redefined accordingly). 

Hope this helps. If not please let me know and I will upload the code for the classifier.

**Comment by cfpark00** [comment]
Thank you! I think I did get something similar! :)

**Comment by heydarshahi** [comment]
Sorry for the repetition; I'm having the exact same problem. What I've tried:

- Normalized the MNIST images to unit vectors from the beginning, by dividing each row by its norm (using the `np.linalg.norm()` function).
- Used [0, 1]-ranged images for the unsupervised part, but normalized the images (to unit vectors) to get the 1st layer activations, then fed them to the gradient descent to train the second layer using Adam optimizer.

Both approaches give me ~94% validation accuracy...

I'd appreciate any feedback on where I might be going wrong.


Many thanks! 

Amin

--
More details:

I've kept all of the hyperparameters the same as the original notebook. However, I used `Keras` to load MNIST  and kept 50000 images for training and 10000 for validation. 

Also, I've refactored the _forward_ computation of the input currents so that I can call it after the unsupervised part is converged. In other words, after convergence, I normalize the MNIST vectors to unit vectors and use that as input to my forward function.
